riscv/src/continuations/bootloader.rs

```diff
@@ -31,6 +32,199 @@ pub const MEMORY_HASH_START_INDEX: usize = 2 * REGISTER_NAMES.len();
+    }
+
+    pub fn final_memory_hash(&self) -> &[B::Fe] {
+        let hash_start = (MEMORY_HASH_START_INDEX + WORDS_PER_HASH) * B::FE_PER_WORD;
+        &self.0[hash_start..hash_start + WORDS_PER_HASH * B::FE_PER_WORD]
+    }
+
+    fn proof_start(input_page_idx: usize) -> usize {
+        B::FE_PER_WORD
+            * (PAGE_INPUTS_OFFSET
+                + BOOTLOADER_INPUTS_PER_PAGE * input_page_idx
+                + 1
+                + WORDS_PER_PAGE
+                + WORDS_PER_HASH)
+    }
+
+    pub fn update_proof(&mut self, input_page_idx: usize, sibling_idx: usize, hash: &[B::Fe]) {
+        assert_eq!(hash.len(), WORDS_PER_HASH * B::FE_PER_WORD);
+        let sibling_start =
+            Self::proof_start(input_page_idx) + sibling_idx * WORDS_PER_HASH * B::FE_PER_WORD;
+        let sibling_end =
+            Self::proof_start(input_page_idx) + (sibling_idx + 1) * WORDS_PER_HASH * B::FE_PER_WORD;
+        self.0[sibling_start..sibling_end].copy_from_slice(hash);
+    }
+
+    pub fn proof(&self, input_page_idx: usize, sibling_idx: usize) -> &[B::Fe] {
+        let sibling_start =
+            Self::proof_start(input_page_idx) + sibling_idx * WORDS_PER_HASH * B::FE_PER_WORD;
+        let sibling_end =
+            Self::proof_start(input_page_idx) + (sibling_idx + 1) * WORDS_PER_HASH * B::FE_PER_WORD;
+        &self.0[sibling_start..sibling_end]
+    }
+
+    pub fn update_page_hash(&mut self, input_page_idx: usize, hash: &[B::Fe]) {
+        assert_eq!(hash.len(), WORDS_PER_HASH * B::FE_PER_WORD);
+        let hash_start =
+            (PAGE_INPUTS_OFFSET + BOOTLOADER_INPUTS_PER_PAGE * input_page_idx + 1 + WORDS_PER_PAGE)
+                * B::FE_PER_WORD;
+        self.0[hash_start..hash_start + WORDS_PER_HASH * B::FE_PER_WORD].copy_from_slice(hash);
+    }
+
+    pub fn update_final_registers(&mut self, register_values: &[B::Fe]) {
+        assert_eq!(register_values.len(), REGISTER_NAMES.len() * B::FE_PER_WORD);
+        let final_registers_start = REGISTER_NAMES.len() * B::FE_PER_WORD;
+        self.0
+            [final_registers_start..final_registers_start + REGISTER_NAMES.len() * B::FE_PER_WORD]
+            .copy_from_slice(register_values);
+    }
+
+    pub fn update_final_root(&mut self, hash: &[B::Fe]) {
+        assert_eq!(hash.len(), WORDS_PER_HASH * B::FE_PER_WORD);
+        let root_hash_start = (MEMORY_HASH_START_INDEX + WORDS_PER_HASH) * B::FE_PER_WORD;
+        self.0[root_hash_start..root_hash_start + WORDS_PER_HASH * B::FE_PER_WORD]
+            .copy_from_slice(hash);
+    }
+
+    pub fn pc(&mut self) -> B::Fe {
+        // the PC is a powdr asm register, which is a single fe, but for the
+        // bootloader inputs, we store it as a "word", the same representation
+        // as the other registers in memory, which may be more than one fe
+        // (e.g., 2 for small fields such as BabyBear).
+        // Here we return the composed fe value.
```

Comment:
I think we could keep the pc as a single FE because in asm we do that right now.

Reply:
I think its better to keep everything as 2 fe in the bootloader itself, because it all becomes "bootloader inputs", which are two columns in the small field machine.